### PR TITLE
Add kubeconfig access handling

### PR DESCRIPTION
**Pull request by m00g3n**
**Description**

Changes proposed in this pull request:

- Add kubeconfig access handling

**Related issue(s)**

See also:
- https://github.com/kyma-project/infrastructure-manager/issues/198

### CODE REVIEW

**Review by koala7659 — COMMENTED** [review]
Please add code to update status when the function is successful

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+		removed = append(removed, crb)
+	}
+
+	return removed
+}
+
+//nolint:gochecknoglobals
+var newContainsAdmin = func(admin string) func(rbacv1.ClusterRoleBinding) bool {
+	return func(r rbacv1.ClusterRoleBinding) bool {
+		for _, subject := range r.Subjects {
```
Comment (koala7659) [inline internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:117]: As we create only ClusterRoleBinding having single subject, we do not need iterate over subjects here. This should be enough: 

```
		if len(r.Subjects) != 1 {
			return false
		}
```

Comment (koala7659) [inline internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:117]: You should also add some checking if given ClusterRole Binding has correct labels:

		if !labels.Set(r.Labels).AsSelector().Matches(labels.Set(labelsClusterRoleBindings)) {
			// cluster role binding is not controlled by KIM
			return false
		}

**Review by koala7659 — CHANGES_REQUESTED** [review]
Please see my comments

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+	}
+)
+
+func sFnApplyClusterRoleBindings(ctx context.Context, m *fsm, s *systemState) (stateFn, *ctrl.Result, error) {
+	// prepare subresource client to request admin kubeconfig
+	srscClient := m.ShootClient.SubResource("adminkubeconfig")
+	shootAdminClient, err := GetShootClient(ctx, srscClient, s.shoot)
+	if err != nil {
+		updateCRBApplyFailed(&s.instance)
+		return updateStatusAndStopWithError(err)
```
Comment (koala7659) [inline internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:32]: Are we OK with returning error from reconcile function?

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+		return nil
+	}
+}
+
+func updateCRBApplyFailed(rt *imv1.Runtime) {
+	rt.UpdateStatePending(
+		imv1.ConditionTypeRuntimeConfigured,
+		imv1.ConditionReasonConfigurationErr,
+		string(metav1.ConditionFalse),
+		"failed to update kubeconfig admin access",
```
Comment (koala7659) [inline internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:190]: The error message is too generic, suggest to pass it as the parameter to make it more specific.

Comment (m00g3n) [reply internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:190]: The detailed message is visible in the logs of the controller, we need to make sure we will not leak or give to much details to confuse the user.

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+	for _, fn := range []func() error{
+		newDelCRBs(ctx, shootAdminClient, removed),
+		newAddCRBs(ctx, shootAdminClient, missing),
+	} {
+		if err := fn(); err != nil {
+			updateCRBApplyFailed(&s.instance)
+			return updateStatusAndStopWithError(err)
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (m00g3n) [reply internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:53]: Can we do in in the separate PR?

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+		missing = append(missing, crb)
+	}
+
+	return missing
+}
+
+func toAdminClusterRoleBinding(name string) rbacv1.ClusterRoleBinding {
+	return rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "admin-",
```
[the remark this answers is not in the document]
Comment (m00g3n) [reply internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:145]: Do we need to change it? Do we expect CRB to have specific name?

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go (hunk 1 above)
Comment (m00g3n) [reply internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:117]: Yes it's true but why should we exclude scenario with multiple subject?
We cannot also assume the subject list will always be unmodified.

Comment (m00g3n) [reply internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:117]: We are checking those labels:
```
	labelsClusterRoleBindings = map[string]string{
		"app":                                   "kyma",
		"reconciler.kyma-project.io/managed-by": "kim",
	}
```

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go (hunk 2 above)
Comment (m00g3n) [reply internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:32]: It's the same as if we would log error and reque:
- https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile#Reconciler

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return shootClientWithAdmin, nil
+}
+
+func isRBACUserKind(s rbacv1.Subject) bool {
+	return s.Kind == rbacv1.UserKind &&
+		s.APIGroup == rbacv1.GroupName
```
[the remark this answers is not in the document]
Comment (m00g3n) [reply internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:86]: To be on the same page do you want just to check if the subject is `User`?

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+func getRemoved(crbs []rbacv1.ClusterRoleBinding, admins []string) (removed []rbacv1.ClusterRoleBinding) {
+	// iterate over cluster role bindings to find out removed administrators
+	for _, crb := range crbs {
+		if !labels.Set(crb.Labels).AsSelector().Matches(labels.Set(labelsClusterRoleBindings)) {
+			// cluster role binding is not controlled by KIM
+			continue
+		}
+
+		index := slices.IndexFunc(crb.Subjects, isRBACUserKind)
```
[the remark this answers is not in the document]
Comment (m00g3n) [reply internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:97]: - https://github.com/kyma-project/infrastructure-manager/pull/323#discussion_r1714931699

File: internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (m00g3n) [reply internal/controller/runtime/fsm/runtime_fsm_apply_clusterrolebindings.go:53]: As we agreed offline we no longer need workaround with the `isio-system` namespace.

**Review by koala7659 — APPROVED** [review]
LGTM

**Pull request merged**